### PR TITLE
feat(gce): document how to deploy Pro Kubernetes cluster on GCE

### PR DESCRIPTION
**Pull request by rebornplusplus**
Document how to deploy an Ubuntu Pro Kubernetes cluster on GKE.

### CODE REVIEW

**Comment by rebornplusplus** [comment]
Preview of `/google/gke/gke-how-to/deploy-pro-gke-cluster.html`:

image

**Comment by rebornplusplus** [comment]
@cjdcordeiro will you please take another look?

**Comment by k-dimple** [comment]
> @cjdcordeiro will you please take another look?

@cjdcordeiro and @rebornplusplus: The content can be viewed at https://canonical-public-cloud--28.com.readthedocs.build/en/28/google/gke/gke-how-to/deploy-pro-gke-cluster/

**Comment by cjdcordeiro** [comment]
> > @cjdcordeiro will you please take another look?
> 
> @cjdcordeiro and @rebornplusplus: The content can be viewed at https://canonical-public-cloud--28.com.readthedocs.build/en/28/google/gke/gke-how-to/deploy-pro-gke-cluster/

neat trick

File: google/gke/gke-how-to/deploy-pro-gke-cluster.rst [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+
+	"Modifications on the boot disk of a node VM do not persist across node re-creations.
+	Nodes are re-created during manual upgrade, auto-upgrade, auto-repair, and auto-scaling.
+	In addition, nodes are re-created when you enable a feature that requires node re-creation,
+	such as GKE Sandbox, intranode visibility, and shielded nodes."
+
+	-- `GKE docs <https://cloud.google.com/kubernetes-engine/docs/concepts/node-images#modifications>`_
+
+Given that, currently, there's no mechanism to either enable Ubuntu Pro or pre-bake the UA token
+in the specific cluster, so the managed Kubernetes is not applicable for the GCE.
```
Comment (cjdcordeiro) [inline google/gke/gke-how-to/deploy-pro-gke-cluster.rst:20]:
```suggestion
in the specific cluster, so the managed Pro Kubernetes cluster is not applicable in GCE.
```

File: google/gke/gke-how-to/deploy-pro-gke-cluster.rst [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+
+Deploy Kubernetes manually on Ubuntu Pro VMs
+--------------------------------------------
+
+One solution to have a Ubuntu Pro cluster is to deploy and manage Kubernetes manually on
```
Comment (cjdcordeiro) [inline google/gke/gke-how-to/deploy-pro-gke-cluster.rst:26]:
```suggestion
One solution to have an Ubuntu Pro cluster is to deploy and manage Kubernetes manually on
```

File: google/gke/gke-how-to/deploy-pro-gke-cluster.rst [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+Ubuntu Pro VMs available in GCP.
```
Comment (cjdcordeiro) [inline google/gke/gke-how-to/deploy-pro-gke-cluster.rst:27]:
```suggestion
Ubuntu Pro VMs available in GCE.
```

GCP was correct, but I think GCE makes as much sense here and it avoids using too many G** acronyms

File: google/gke/gke-how-to/deploy-pro-gke-cluster.rst [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+Create Ubuntu Pro VMs for the control plane and worker nodes
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Create a few Ubuntu Pro VMs for the worker nodes and control plane from the
+`GCE VM Instances <https://console.cloud.google.com/compute/instances>`_ page.
+There is a `Create Instance <https://console.cloud.google.com/compute/instancesAdd>`_ button
+which takes you to the configuration details before creating the VM.
+
+Make sure to change the boot disk configuration so that the operating system
+is "Ubuntu Pro". There are a few versions to choose from:
```
Comment (cjdcordeiro) [inline google/gke/gke-how-to/deploy-pro-gke-cluster.rst:38]:
```suggestion
is "Ubuntu Pro". There are a few versions to choose from, such as:
```

(there could be more in the future)

File: google/gke/gke-how-to/deploy-pro-gke-cluster.rst [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+You can also utilise the `gcloud CLI tool <https://cloud.google.com/sdk/gcloud>`_ to deploy the VMs
+with the following command.
+
+::
+
+	$ gcloud compute instances create <instance-name> <options..>
+
+As an example, you can create two Ubuntu Pro VMs for the worker nodes named ``k8s-worker-1``
+and ``k8s-worker-2`` and another Pro VM for the control plane named ``k8s-main``, each of them
+on "Ubuntu Pro" operating system with "Ubuntu 20.04 LTS Pro FIPS Server" version.
```
Comment (cjdcordeiro) [inline google/gke/gke-how-to/deploy-pro-gke-cluster.rst:58]:
```suggestion
```

File: google/gke/gke-how-to/deploy-pro-gke-cluster.rst [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+	$ gcloud compute ssh --zone <instance-zone> <instance-name> --project <project-name>
+
+Install Kubernetes and create a cluster
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Install Kubernetes on the nodes to create and manage the cluster. You can use
+`MicroK8s <https://microk8s.io/>`_ to meet your Kubernetes needs. Install the snap.
+
+::
+
```
Comment (cjdcordeiro) [inline google/gke/gke-how-to/deploy-pro-gke-cluster.rst:73]:
```suggestion

	# repeat for each node
```

File: google/gke/gke-how-to/deploy-pro-gke-cluster.rst [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+	$ sudo snap install microk8s --classic
+
+To create a cluster out of two or more already-running MicroK8s instances, use the
+``microk8s add-node`` command. The MicroK8s instance on which this command is run
+will be the master of the cluster and will host the Kubernetes control plane.
```
Comment (cjdcordeiro) [inline google/gke/gke-how-to/deploy-pro-gke-cluster.rst:78]:
```suggestion
will be the cluster's manager and will host the Kubernetes control plane.
```

File: google/gke/gke-how-to/deploy-pro-gke-cluster.rst [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+
+Following the example VMs, run the command in ``k8s-main``.
```
Comment (cjdcordeiro) [inline google/gke/gke-how-to/deploy-pro-gke-cluster.rst:80]:
```suggestion
Following the exemplary VMs from above, run the following command in ``k8s-main``.
```

File: google/gke/gke-how-to/deploy-pro-gke-cluster.rst [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+	microk8s join 10.128.0.7:25000/3a9547446eae54765ecc0bc522321968/079215076aa1 --worker
+
+	If the node you are adding is not reachable through the default interface you can use one of the following:
+	microk8s join 10.128.0.7:25000/3a9547446eae54765ecc0bc522321968/079215076aa1
+
+Note that, the IP address, port, and token generated here will be project specific.
+You will not have the same values as above. Find more info in the
+`microk8s clustering doc <https://microk8s.io/docs/clustering>`_.
+
+In the worker nodes, execute the respective command to join the cluster as a worker.
```
Comment (cjdcordeiro) [inline google/gke/gke-how-to/deploy-pro-gke-cluster.rst:98]:
```suggestion
Then, SSH into each of the worker nodes and execute the respective command to join the cluster as a worker.
```

File: google/gke/gke-how-to/deploy-pro-gke-cluster.rst [hunk 10]
```diff
@@ -0,0 +1,137 @@
...
+
+::
+
```
Comment (cjdcordeiro) [inline google/gke/gke-how-to/deploy-pro-gke-cluster.rst:101]:
```suggestion

	# repeat for each worker node
```

File: google/gke/gke-how-to/deploy-pro-gke-cluster.rst [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+	root@k8s-main:~# microk8s kubectl cluster-info
+	Kubernetes control plane is running at https://127.0.0.1:16443
+
+	To further debug and diagnose cluster problems, use 'kubectl cluster-info dump'.
+
+
+Conclusion
+~~~~~~~~~~
+
+You now have a Ubuntu Pro Kubernetes cluster running on Ubuntu Pro nodes in GCP.
```
Comment (cjdcordeiro) [inline google/gke/gke-how-to/deploy-pro-gke-cluster.rst:137]: Can you please add a one-liner to show **how** we know we are running in a Pro cluster? I guess you can use either the `ua` or `pro` CLI clients to assert that. So something like:

```suggestion
You now have an Ubuntu Pro Kubernetes cluster running in GCE. Your Pro subscription can be verified on each of the provisioned nodes by running:

#### TODO: 

And, from the control plane node (``k8s-main`` in the example above), you'll also find the _KUBECONFIG_ files at ``TODO``, which you can then use to control your Pro Kubernetes cluster from any working environment with a Kubernetes client.
```

**Review by cjdcordeiro — CHANGES_REQUESTED** [review]
Looks good! Thanks.

I have a few nitpicks and a minor request for change in the last comment.

I also have 1 question: there is no reference about firewalls nor networking configuration, so I'm pretty sure you wouldn't be able to access these VMs from the outside (as you'll probably confirm when you write the "How to deploy a Pro container image"). The VMs will need to be granted network access on the ports used by the control plane, such that you can access the cluster from your workstation. Can you please add a few lines about this (maybe next to the VM deployment instructions).

File: google/gke/gke-how-to/deploy-pro-gke-cluster.rst (hunk 11 above)
Comment (rebornplusplus) [reply google/gke/gke-how-to/deploy-pro-gke-cluster.rst]: Done.

**Comment by rebornplusplus** [comment]
Hi @AndrewCloke and @cjdcordeiro, thanks for your reviews! I have updated the name, file location, and the contents accordingly. Please take another look and let me know what you think!

**Comment by cjdcordeiro** [comment]
Nice. LGTM

File: google/gce/gce-how-to/deploy-kubernetes-with-ubuntu-pro.rst [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+
+	"Modifications on the boot disk of a node VM do not persist across node re-creations.
+	Nodes are re-created during manual upgrade, auto-upgrade, auto-repair, and auto-scaling.
+	In addition, nodes are re-created when you enable a feature that requires node re-creation,
+	such as GKE Sandbox, intranode visibility, and shielded nodes."
+
+	-- `GKE docs <https://cloud.google.com/kubernetes-engine/docs/concepts/node-images#modifications>`_
+
+Given that, currently, there's no mechanism to either enable Ubuntu Pro or pre-bake the UA token
+in the specific cluster, so the managed Pro Kubernetes cluster is not applicable in GCE.
```
Comment (philroche) [inline google/gce/gce-how-to/deploy-kubernetes-with-ubuntu-pro.rst:20]: > Kubernetes cluster is not applicable in GCE.

Shouldn't this say GKE?

File: google/gce/gce-how-to/deploy-kubernetes-with-ubuntu-pro.rst [hunk 13]
```diff
@@ -0,0 +1,244 @@
+Deploy Kubernetes with Ubuntu Pro on GCE
+========================================
+
+
+Limitations
+-----------
+
+Google does not yet have Ubuntu Pro image offerings for the GKE nodes i.e. you cannot choose
```
Comment (philroche) [inline google/gce/gce-how-to/deploy-kubernetes-with-ubuntu-pro.rst:8]: People might not know what GKE is

> image offerings for the GKE nodes

image offerings for the GKE (Google Kubernetes Engine) nodes

**Review by philroche — CHANGES_REQUESTED** [review]

File: google/gce/gce-how-to/deploy-kubernetes-with-ubuntu-pro.rst (hunk 12 above)
Comment (rebornplusplus) [reply google/gce/gce-how-to/deploy-kubernetes-with-ubuntu-pro.rst]: Thanks. Made a typo I think.

File: google/gce/gce-how-to/deploy-kubernetes-with-ubuntu-pro.rst (hunk 13 above)
Comment (rebornplusplus) [reply google/gce/gce-how-to/deploy-kubernetes-with-ubuntu-pro.rst]: Added a link as well! Thank you.

**Review by philroche — APPROVED** [review]

**Comment by rebornplusplus** [comment]
Rebased, this can be merged now!

**Pull request merged**